### PR TITLE
First version of new stream/signal library

### PR DESCRIPTION
**Pull request by jvanbruegge**
**Do not merge yet**

Just a new API for the stream/signal lib. Currently without "fluff"

### CODE REVIEW

**Comment by jvanbruegge** [comment]
@staltz An implementation like [here](https://github.com/staltz/experiment-push-pull-cyclejs/pull/3/files#diff-977390b0aad26a16549ee97a7ded446bR37) is not efficient enough, is it?
I'm not really experienced with performance in Javascript.

**Comment by staltz** [comment]
@jvanbruegge cool that you're experimenting with this :)
I took a look, but could you summarize what is your goal when designing the API?

**Comment by jvanbruegge** [comment]
Basicly have one primitive with the core of a signal/observable - an object with a next function for signals and an object with a subscribe function for observables - and a bunch of (at least from the outside) pure functions tranforming those primitives.

To allow chaining, the primitives have the pure functions as members, with `this` bound as first argument.

**Comment by jvanbruegge** [comment]
[This](https://github.com/staltz/experiment-push-pull-cyclejs/pull/3/commits/9505a8dd005b9766e861149a5dd4577cda6ae238#diff-977390b0aad26a16549ee97a7ded446bR42) is what I want. Curried functions I can compose once and store in a constant (e.g. with ramda `compose`/`pipe`) and all those functions also available on the primitive for easy chaining

**Comment by staltz** [comment]
When you say "signal/observable", do you mean one object that satisfies *both* contracts or two objects one for each contract?

**Comment by jvanbruegge** [comment]
one for each
One is push, one is pull, Would not make sense to put them both on the same object. Also having them seperated lead to clearer semantics, like `sampleCombine`, taking one _stream_ and n _signals_ to create one _stream_

**Comment by jvanbruegge** [comment]
The new API would allow for something like this:
```ts
const second = nth(1);
const sample = signal => compose(map(second), sampleCombine(signal))

const currentRandom$ = click$.compose(sample(randomS));
```

**Comment by jvanbruegge** [comment]
I wasn't sure about having multiple libraries, but I kinda like it.
From the outside, the API does not have changed much, you just have a bunch of new named exports that can be used freely.

**Comment by staltz** [comment]
I took a look at your zCombinators idea. It's useful because xs can't import ys and ys import xs, that would be a circular dependency between npm modules, and it's complicated to ask in which library should some operator live. But then I also noticed that since `Stream implements Observable` and `Signal implements Iterable`, we can put "signal+stream" operators in xstream as long as the expected signal is type `Iterable`, and we can put "signal+stream" operators in ysignal as long as the expected stream is type `Observable`. Which operator goes to which library? I thought maybe "if it returns type A, it should be in library A" would be a good rule of thumb.

**Comment by jvanbruegge** [comment]
i would add it to the library that the input is. Like `startWith` taking an stream and returning a signal. It would put that operator in xs, so you can add it on the stream class.

But im not entirely sure we should do this, as Iterator and signal is not the same. A signal is an infinite iterator. So passing an interator like array would not throw an error in typescript

**Comment by jvanbruegge** [comment]
I have most of xstreams API, currently missing is `unsubscribe()`, error handling and it is not "hot".

**Comment by staltz** [comment]
>But im not entirely sure we should do this, as Iterator and signal is not the same. A signal is an infinite iterator.

Actually `Signal implements Iterable` (not Iterator). Just like a Subscription represents the execution of an Observable, an Iterator represents the execution of an Iterable.

> i would add it to the library that the input is. Like startWith taking an stream and returning a signal. It would put that operator in xs, so you can add it on the stream class.

It think this is a complex topic. (Definition: "collection" is a stream or a signal) If there is one collection as input and one collection as output, (e.g. startWith has two inputs: stream `this` and arg `t: T`, but one output collection) we can either model that as an operator on the input (e.g. `stream.startWith`) or as a factory of the output type (e.g. `ys.fromStream`). I think we should treat `Signal -> Stream` and `Stream -> Signal` as being explicitly conversion factories (e.g. `fromSomething`). When crossing the push<-->pull boundary, some information loss happens (refer to [this part of the big Cycle.js Push Pull discussion issue](https://github.com/cyclejs/cyclejs/issues/581#issuecomment-302163876)) and we might need additional parameters to specify how to convert, so it might be `Stream -> a -> Signal` but as a factory, not an operator (in Haskell it doesn't matter, but in OOP chaining style it matters): `fromStream : <T>(stream: Stream<T>, t: T) => Signal<T>`. Then, there are operators that go from collection type `A` to type `A`, but use another collection type in between, like `sampleCombine : <T,R>(this: Stream<T>, sig: Signal<R>) => Stream<[T,R]>` and that one could live in a zCombinator library or just in `xs` library where we have the signature `sampleCombine : <T,R>(this: Stream<T>, sig: Iterable<R>) => Stream<[T,R]>` instead.

**Comment by jvanbruegge** [comment]
> When crossing the push<-->pull boundary, some information loss happens 

Of course, push -> pull discards the "when did it happen" aspect
pull -> push (something like a trigger) would require polling, not sure if we should implement that

> but as a factory, not an operator

well, with the new API, every operator is a factory, the Stream class only has those factories `composed` with itself
